### PR TITLE
add prune_layer support to GrowingModule

### LINKED ISSUE #257: Add neuron/channel pruning support to GrowingModule (prune_layer + prune_layer_in/out)
## Motivation

gromo can grow the hidden dimension of a layer (`create_layer_extensions` /
`extended_*_layer`) but has no symmetric operation to **shrink** it. Pruning is
needed for experiments that combine growing with compression — e.g. warm-start
from a trained checkpoint, prune to a target parameter budget, then re-grow.

## Proposed change

Add a symmetric counterpart to layer extension:

- `GrowingModule.prune_layer(indices_to_remove)` — high-level entry point that
  removes hidden-dim neurons between `previous_module` and `self`. It:
  1. validates indices (empty = no-op, out-of-range = `IndexError`, cannot
     prune all neurons),
  2. drops the matching output rows of `previous_module` via `prune_layer_out`,
  3. shrinks any `post_layer_function` with `num_features` (e.g. BatchNorm:
     weight/bias/running_mean/running_var),
  4. drops the matching input columns of `self` via `prune_layer_in`,
  5. resets `tensor_s` / `previous_tensor_{s,m}` since they were sized for the
     old hidden dim.

- `LinearGrowingModule.prune_layer_in / prune_layer_out` — rebuild `self.layer`
  as a smaller `nn.Linear` keeping the surviving columns / rows.

- `Conv2dGrowingModule.prune_layer_in / prune_layer_out` — rebuild `self.layer`
  as a smaller `nn.Conv2d` keeping the surviving in- / out-channels (preserving
  stride, padding, dilation, groups, padding_mode).

## Out of scope

- Pruning *criteria* (which neurons to drop) — that lives in the experiment
  layer, not in gromo. This PR only provides the mechanism.

### PR DESCRIPTION
**Pull request by PineappleBlowsnow**
## Summary

Adds neuron/channel pruning to gromo — the symmetric counterpart of layer
growing. Lets you shrink the hidden dim between two growing modules.

Closes #257 

## Changes

- `GrowingModule.prune_layer(indices)` — removes hidden-dim neurons: drops
  `previous_module` output rows, shrinks BatchNorm in the post-layer function,
  drops `self` input columns, resets stats.
- `LinearGrowingModule.prune_layer_in/out` — rebuild `self.layer` as a smaller
  `nn.Linear`.
- `Conv2dGrowingModule.prune_layer_in/out` — rebuild `self.layer` as a smaller
  `nn.Conv2d` (preserves stride/padding/dilation/groups).

## Out of scope

Pruning criteria (which neurons to drop) — lives in the experiment layer.

## TODO before merge

- [OK ] Add tests under `tests/` (linear, conv, BatchNorm, edge cases).

### CODE REVIEW

File: src/gromo/modules/growing_module.py [hunk 1]
```diff
@@ -3760,6 +3760,88 @@ def create_layer_extensions(
             noise_ratio=noise_ratio,
         )
 
+    @torch.no_grad()
+    def prune_hidden_neurons_in(
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: I think it does not need to be called "hidden". They will inevitably be hidden but I think there is nothing specific about this and the name might be confusing.

Similarly the docstring is a bit confusing -- it is just pruning the fan-in neurons.

File: src/gromo/modules/growing_module.py [hunk 2]
```diff
@@ -3760,6 +3760,88 @@ def create_layer_extensions(
...
+        # 0) Need a previous GrowingModule to coordinate with.
+        assert isinstance(self.previous_module, GrowingModule), (
+            f"prune_hidden_neurons_in requires a previous GrowingModule, "
+            f"got {type(self.previous_module)}"
+        )
+
+        # 1) Normalize + validate indices.
+        if len(indices_to_remove) == 0:
+            return
+        indices = sorted({int(i) for i in indices_to_remove})
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: For all these functions, I think it's cleaner to use a `numpy` array for the index list. This makes it cleaner and faster to assert the type, check the min/max, invert it, etc.

Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: We should assert that the type is `int`, rather than casting.

File: src/gromo/modules/growing_module.py [hunk 3]
```diff
@@ -3760,6 +3760,88 @@ def create_layer_extensions(
...
+
+        post_fn = self.previous_module.post_layer_function
+        if post_fn is not None:
+            sub_modules = (
+                list(post_fn) if isinstance(post_fn, torch.nn.Sequential) else [post_fn]
+            )
+            for m in sub_modules:
+                if not hasattr(
+                    m, "num_features"
+                ):  # skip modules that don't have num_features (e.g. ReLU)
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: Currently, I don't think this handles e.g. GroupNorm, because that uses `num_channels` instead of `num_features`?

File: src/gromo/modules/growing_module.py [hunk 4]
```diff
@@ -3760,6 +3760,88 @@ def create_layer_extensions(
...
+                        m.running_mean.device
+                        if getattr(m, "running_mean", None) is not None
+                        else "cpu"
+                    )
+                )
+
+                keep_mask = torch.ones(m.num_features, dtype=torch.bool, device=device)
+
+                keep_mask[indices] = False
+                if isinstance(getattr(m, "weight", None), torch.nn.Parameter):
```
Comment (alexdavey) [inline src/gromo/modules/growing_module.py]: We don't really want to have module/function specific code in `growing_module.py`. In addition, this may be unsafe, since we don't have the guarantee that the weights we are pruning are in fact the correct ones since we do not know what operation these weights correlate to.

We can discuss in person if you want, but I think a safer approach is to:
1) Identify if the `post_layer_function`s are activations, e.g. using something like

```
import torch.nn as nn
import torch.nn.modules.activation as activation_mod
import torch.nn.modules.dropout as dropout_mod

def is_passthrough(module: nn.Module) -> bool:
    cls = module.__class__

    return (
        cls.__module__ == activation_mod.__name__
        and len(list(module.parameters(recurse=False))) == 0
        and len(list(module.buffers(recurse=False))) == 0
        and not isinstance(module, (nn.PReLU, nn.MultiheadAttention, nn.GLU, dropout_mod._DropoutNd))
    )
```

2) For anything outside of this list, we only apply pruning to modules in a specific list (e.g. GrowingGroupNorm, GrowingLayerNorm, etc), which all have a `.prune` function attached to it which describes the detailed mechanics of how to prune this module.

This approach should be safer, and avoids putting module-specific code in the main `GrowingModule` class.

File: src/gromo/modules/linear_growing_module.py [hunk 5]
```diff
@@ -1100,3 +1100,76 @@ def create_layer_out_extension(self, extension_size: int) -> None:
...
+        Parameters
+        ----------
+        indices_to_remove: list[int]
+            Column indices in ``self.layer.weight`` to drop. Assumed already
+            de-duplicated, sorted, and in-range by the caller.
+        """
+        keep = [i for i in range(self.in_features) if i not in set(indices_to_remove)]
+        keep_idx = torch.tensor(keep, device=self.device, dtype=torch.long)
+
+        new_layer = torch.nn.Linear(
```
Comment (alexdavey) [inline src/gromo/modules/linear_growing_module.py:1137]: As with the conv, we need to preserve the `requires_grad` of the previous layer. Also for `prune_layer_out`.

File: src/gromo/modules/conv2d_growing_module.py [hunk 6]
```diff
@@ -1236,6 +1236,81 @@ def layer_out_extension(
...
+        )
+
+    @torch.no_grad()
+    def prune_layer_out(self, indices_to_remove: list[int]) -> None:
+        """Shrink "self.layer"by dropping the output channels."""
+        keep = [i for i in range(self.out_features) if i not in set(indices_to_remove)]
+        device = self.layer.weight.device
+        keep_idx = torch.tensor(keep, device=device, dtype=torch.long)
+
+        new_layer = torch.nn.Conv2d(
```
Comment (alexdavey) [inline src/gromo/modules/conv2d_growing_module.py:1326]: We should preserve the `requires_grad` of the original layer. Same for the `prune_layer_in` functions.

File: tests/test_linear_growing_module.py [hunk 7]
```diff
@@ -465,6 +465,158 @@ def setUp(self):
...
+        self.assertEqual(first.out_features, second.in_features)
+        # surviving hidden neurons keep their original weights
+        self.assertAllClose(first.weight, original_first_weight.index_select(0, keep))
+        self.assertAllClose(second.weight, original_second_weight.index_select(1, keep))
+        # chain forward still works and preserves output shape
+        x = torch.randn(self.n, first.in_features, device=global_device())
+        y = second(first(x))
+        self.assertShapeEqual(y, (self.n, second.out_features))
+
+    def test_prune_hidden_neurons_in_with_batchnorm(self):
```
Comment (alexdavey) [inline tests/test_linear_growing_module.py]: We should also add tests for layernorm, groupnorm, GrowingBatchNorm etc.

File: tests/test_conv2d_growing_module.py [hunk 8]
```diff
@@ -677,6 +677,88 @@ def setUp(self):
...
+        layer.tensor_s.init()
+        x = torch.randn(5, layer.in_channels, 8, 8, device=global_device())
+        layer(x)
+        layer.tensor_s.update()  # would trip the size assert if _shape were wrong
+
+        k0, k1 = layer.layer.kernel_size
+        in_dim = layer.in_channels * k0 * k1 + 1  # +1 for bias
+        self.assertEqual(layer.tensor_s.samples, x.size(0))
+        self.assertShapeEqual(layer.tensor_s(), (in_dim, in_dim))
+
```
Comment (alexdavey) [inline tests/test_conv2d_growing_module.py:792]: We should also add, somewhere, a test for BatchNorm (etc) here. Currently we only do these tests for the `linear_growing_module` while `conv` is the one which is actually important.

File: tests/test_linear_growing_module.py [hunk 9]
```diff
@@ -465,6 +465,158 @@ def setUp(self):
...
+        """The previous module's BatchNorm post-layer is pruned alongside the hidden dim."""
+        hidden = 4
+        first, second = self.create_demo_layers(
+            bias=True,
+            hidden_features=hidden,
+            first_layer_post_layer=torch.nn.BatchNorm1d(hidden, device=global_device()),
+        )
+        bn = first.post_layer_function
+        # give every channel a distinct value so we can check the right ones survive
+        with torch.no_grad():
```
Comment (alexdavey) [inline tests/test_linear_growing_module.py:566]: It's a nitpick, but I think we can just use random values here as it's a bit simpler.

File: tests/test_conv2d_growing_module.py [hunk 10]
```diff
@@ -677,6 +677,88 @@ def setUp(self):
 
         self.demo_couple = {b: self.create_demo_layers(b) for b in (True, False)}
 
+    # ------------------------------------------------------------------ #
```
Comment (alexdavey) [inline tests/test_conv2d_growing_module.py:681]: Where possible, we should instead call these tests in `test_growing_module` on the top-level function, rather than the internal ones.

**Comment by alexdavey** [comment]
Thank you for the PR @PineappleBlowsnow ! I have made some comments.

File: src/gromo/modules/conv2d_growing_module.py (hunk 6 above)
Comment (PineappleBlowsnow) [reply src/gromo/modules/conv2d_growing_module.py:1326]: Done
